crates/qpy/src/circuits.rs

```diff
@@ -0,0 +1,2166 @@
+    };
+
+    let annotation_deserializers_data: Vec<(String, Bytes)> = packed_circuit
+        .annotation_headers
+        .state_headers
+        .iter()
+        .map(|data| (data.namespace.clone(), data.state.clone()))
+        .collect();
+    qpy_data
+        .annotation_handler
+        .load_deserializers(annotation_deserializers_data)?;
+    let global_phase = packed_circuit
+        .header
+        .global_phase_data
+        .to_param(py, &mut qpy_data)?;
+    qpy_data.circuit_data.set_global_phase(global_phase)?;
+
+    add_standalone_vars(py, &packed_circuit, &mut qpy_data)?;
+    add_registers_and_bits(&packed_circuit, &mut qpy_data)?;
+
+    let custom_instructions = read_custom_instructions(py, &packed_circuit, &mut qpy_data)?;
+    for instruction in &packed_circuit.instructions {
+        let inst = unpack_instruction(py, instruction, &custom_instructions, &mut qpy_data)?;
+        qpy_data.circuit_data.push(inst)?;
+    }
+
+    for (vector, initialized_params) in qpy_data.vectors.values() {
+        let vector_length = vector
+            .bind(py)
+            .call_method0("__len__")?
+            .extract::<usize>()?;
+        let missing_indices: Vec<u64> = (0..vector_length as u64)
+            .filter(|x| !initialized_params.contains(x))
+            .collect();
+        if initialized_params.len() != vector_length {
+            let msg = format!(
+                "The ParameterVector: '{:}' is not fully identical to its \
+                pre-serialization state. Elements {:} \
+                in the ParameterVector will be not equal to the pre-serialized ParameterVector \
+                as they weren't used in the circuit: {:}",
+                vector.getattr(py, "name")?.extract::<String>(py)?,
+                missing_indices
+                    .iter()
+                    .map(|index| index.to_string())
+                    .collect::<Vec<_>>()
+                    .join(", "),
+                packed_circuit.header.circuit_name
+            );
+            imports::WARNINGS_WARN.get_bound(py).call1((
+                msg,
+                imports::BUILTIN_USER_WARNING.get_bound(py),
+                1,
+            ))?;
+        }
+    }
+
+    let unpacked_layout = unpack_layout(py, &packed_circuit.layout, &circuit_data)?;
+    let metadata =
+        deserialize_metadata(py, &packed_circuit.header.metadata, metadata_deserializer)?;
+    let circuit = imports::QUANTUM_CIRCUIT
+        .get_bound(py)
+        .call_method1(intern!(py, "_from_circuit_data"), (circuit_data,))?;
+    // add registers
+
+    circuit.setattr("metadata", metadata)?;
+    circuit.setattr("name", &packed_circuit.header.circuit_name)?;
+    if let Some(layout) = unpacked_layout {
+        circuit.setattr("_layout", layout)?;
+    }
+    Ok((circuit, pos))
+}
+
+#[pyfunction]
+#[pyo3(signature = (file_obj, circuit, metadata_serializer, use_symengine, version, annotation_factories))]
+pub fn py_write_circuit(
```

Comment:
The convention is to not expose the `py_` prefix to Python. So I'd change the Python-side name using the `name` argument of `pyo3`